pkg/networkservice/mechanisms/directmemif/server.go

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package directmemif provides server chain element that create connection between two memif interfaces
+package directmemif
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/serialize"
+	"github.com/pkg/errors"
+
+	"github.com/networkservicemesh/sdk-vppagent/pkg/networkservice/mechanisms/directmemif/proxy"
+	"github.com/networkservicemesh/sdk-vppagent/pkg/networkservice/vppagent"
+)
+
+type directMemifServer struct {
+	executor serialize.Executor
+	proxies  map[string]proxy.Proxy
+}
+
+// NewServer creates new direct memif server
+func NewServer() networkservice.NetworkServiceServer {
+	return &directMemifServer{
+		executor: serialize.NewExecutor(),
+		proxies:  map[string]proxy.Proxy{},
+	}
+}
+
+func (d *directMemifServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
+	c := vppagent.Config(ctx)
+	l := len(c.GetVppConfig().GetInterfaces())
+	if l < 2 {
+		return next.Server(ctx).Request(ctx, request)
+	}
+	client := c.GetVppConfig().GetInterfaces()[l-2]
+	endpoint := c.GetVppConfig().GetInterfaces()[l-1]
+	c.GetVppConfig().Interfaces = c.GetVppConfig().GetInterfaces()[:l-2]
+	if client.GetMemif() == nil || endpoint.GetMemif() == nil {
+		return next.Server(ctx).Request(ctx, request)
+	}
+	var err error
+	d.executor.SyncExec(func() {
+		_, ok := d.proxies[request.Connection.Id]
+		if ok {
+			err = errors.New(fmt.Sprintf("proxy with id %v already exsist", request.Connection.Id))
+		}
+	})
+	if err != nil {
+		return nil, err
+	}
+	p, err := proxy.New(client.GetMemif().GetSocketFilename(), endpoint.GetMemif().GetSocketFilename(), proxy.StopListenerAdapter(func() {
+		d.executor.AsyncExec(func() {
+			delete(d.proxies, request.Connection.Id)
+		})
+	}))
+
+	if err != nil {
+		return nil, err
+	}
+	err = p.Start()
+	if err != nil {
+		return nil, err
+	}
+	return request.Connection, nil
```

Comment:
Don't we need to call next.Server(ctx).Request(...) here?  Or am I confused :)

Reply:
Why directmemif can not be the last of the chain? It looks for me if it found two memif interfaces then we need just create a proxy for them and return connection.

Reply:
directmemif may or may not be last in the chain, we can't know apriori, and we should work correctly even if it isn't.  The basic contract is:

If we the last two interfaces in the VppConfig are memif, we will create a proxy for that case, and remove them from the config.  Anything else going on after that isn't our problem, so we just keep going down the chain :)

Reply:
finally fixed